File: coauthor/agent/model_handler_google.py
"""Handler for Google models using OpenAI-compatible API."""

from typing import Any
from openai import OpenAI

from ..logger import logger

from .model_handler_openai import OpenAIHandler
from .response_usage import OpenAIAPIResponseUsage


class GoogleviaOpenAIHandler(OpenAIHandler):
    """Handler for Google models using OpenAI-compatible API."""

    def getClient(self) -> OpenAI:
        """Get OpenAI client with Google's base URL."""
        api_key = self.get_api_key()
        base_url = self.get_base_url()
        logger.info(f"Using Google API key. Base URL: {base_url}")
        return OpenAI(
            api_key=api_key,
            base_url=base_url,
        )

    def compute_price(self, responseUsage: Any) -> float:
        """Compute price for Google token usage."""
        # Google models return completionTokens, promptTokens instead of completion_tokens, prompt_tokens
        prompt_tokens = getattr(responseUsage, "promptTokens", 0)
        completion_tokens = getattr(responseUsage, "completionTokens", 0)

        return (prompt_tokens * self.config.inputPrice + completion_tokens * self.config.outputPrice) / 1e6

    def computeResponseUsage(self, responseUsage: Any, responseTime: float) -> OpenAIAPIResponseUsage:
        """Compute statistics for Google models."""
        # Create a minimal usage object with Google's token counts
        usage_obj = type(
            "GoogleUsage",
            (),
            {
                "prompt_tokens": getattr(responseUsage, "promptTokens", 0),
                "completion_tokens": getattr(responseUsage, "completionTokens", 0),
                "total_tokens": getattr(responseUsage, "totalTokens", 0),
                "prompt_tokens_details": type("Details", (), {"cached_tokens": 0})(),
                "completion_tokens_details": type(
                    "Details", (), {"reasoning_tokens": 0, "accepted_prediction_tokens": None, "rejected_prediction_tokens": None}
                )(),
            },
        )()

        return OpenAIAPIResponseUsage.from_response(usage_obj, self.compute_price(responseUsage), responseTime)

    def should_continue(self, stopReason: str, newResponse: str, agentSetting) -> bool:
        """Determine if OpenAI model should continue generating."""
        logger.info("Determining if should continue for Google model via OpenAI API")
        return not agentSetting.has_endTag(newResponse)

File: coauthor-for-vs-code/src/agent/modelHandlerGoogle.ts
// Standard library imports
// (none needed)

// Third-party imports
import OpenAI from 'openai';

// Local imports - agent components
import { ModelHandlerOpenAI } from './modelHandlerOpenAI';
import { hasEndTag } from './AgentDataclass';
import { OpenAIAPIResponseUsage, ResponseUsageFactory } from './ResponseUsage';

/**
 * Handler for Google models using OpenAI-compatible API.
 */
export class ModelHandlerGoogle extends ModelHandlerOpenAI {
  /** Returns OpenAI client configured with Google's base URL. */
  getClient(): OpenAI {
    const apiKey = this.getApiKey();
    const baseUrl = this.getBaseUrl();
    this.logger.info(`Using Google API key. Base URL: ${baseUrl}`);
    return new OpenAI({
      apiKey,
      baseURL: baseUrl,
    });
  }

  /** Computes cost based on Google's token usage format. */
  computePrice(responseUsage: any): number {
    // Google models return completionTokens, promptTokens instead of completion_tokens, prompt_tokens
    const promptTokens = responseUsage?.promptTokens ?? 0;
    const completionTokens = responseUsage?.completionTokens ?? 0;

    return (
      (promptTokens * this.config.inputPrice +
        completionTokens * this.config.outputPrice) /
      1e6
    );
  }

  /** Creates usage statistics from Google's response format. */
  computeResponseUsage(
    responseUsage: any,
    responseTime: number,
  ): OpenAIAPIResponseUsage {
    // Create a minimal usage object with Google's token counts
    const usageObj = {
      prompt_tokens: responseUsage?.promptTokens ?? 0,
      completion_tokens: responseUsage?.completionTokens ?? 0,
      total_tokens: responseUsage?.totalTokens ?? 0,
      prompt_tokens_details: { cached_tokens: 0 },
      completion_tokens_details: {
        reasoning_tokens: 0,
        accepted_prediction_tokens: null,
        rejected_prediction_tokens: null,
      },
    };

    return ResponseUsageFactory.fromOpenAIResponse(
      usageObj,
      this.computePrice(responseUsage),
      responseTime,
    );
  }

  /** Determines if generation should continue based on response content. */
  shouldContinue(
    stopReason: string,
    newResponse: string,
    agentSetting: any,
  ): boolean {
    this.logger.info(
      'Determining if should continue for Google model via OpenAI API',
    );
    return !hasEndTag(agentSetting, newResponse);
  }
}
